# Annotated executed notebook: training the 33-class classifier on Colab

This is the **real executed run** (outputs are kept exactly as Colab produced them). Only comments and short notes were added; no code line was changed.
The original file is `notebooks/colab_classifier_training_executed.ipynb`.

**result of this run:** 13 epochs (early stopping), about 52 minutes on a Tesla T4. best validation accuracy 93.86% at epoch 8. test set (4,232 images): item accuracy **93.83%**, family accuracy **97.28%**, hazard recall **98.37%** (726 of 738).

## where is each feature coded?

| feature | file (inside `src/edgewaste/`) | main function or class |
|---|---|---|
| 33 classes, 9 families, data sources | `taxonomy.py` | `CLASSES`, `FAMILIES`, `SOURCES` |
| settings from yaml | `config.py` | `Config.load`, `TrainConfig` |
| download, ingest, split | `data/download_kaggle_datasets.py`, `data/ingest_raw_datasets.py`, `data/make_data_splits.py` | `ingest_source`, `main` |
| image loading, augmentation, sampler weights | `data/waste_image_dataset.py` | `build_transforms`, `WasteDataset.sampler_weights` |
| hybrid model (convnext + vit + attention) | `classification/hybrid_model.py` | `AttentionFusion`, `HybridConvNeXtViT` |
| training loop | `classification/train_classifier.py` | `train` |
| evaluation (item, family, hazard) | `classification/evaluate_classifier.py` | `evaluate`, `_family_level`, `_hazard_recall` |
| uncertainty (monte carlo dropout) | `decision_engine/mc_dropout_uncertainty.py` | `mc_dropout_predict` |
| conformal family-set routing | `decision_engine/conformal_routing.py` | `calibrate`, `prediction_sets` |
| object-material cross-check | `decision_engine/object_identity_prior.py` | `apply_identity_prior` |
| final decision | `decision_engine/routing_rules.py` | `decide` |
| contamination index (sensors) | `contamination/oci_model.py`, `contamination/sensor_normalization.py` | `fit_oci_weights`, `compute_oci` |

Notes with **what / where / why** are added under the important cells. Comments inside the code cells are in simple lowercase english.


colab_classifier_training.ipynb - trains the 33-class classifier on a google colab t4 gpu and keeps checkpoints on drive

# Edge-Waste — 33-Class Classifier Training (Colab)

Trains the ConvNeXt+ViT hybrid classifier on the **33-class / 9-family** taxonomy.

**Run every cell in order.** Steps 3 and 6 are hard gates: they stop the notebook if the
code or the data is not what training expects. Both failure modes are otherwise silent —
you would get a finished model trained on the wrong taxonomy and no error to explain it.

The YOLO detector is **not** trained here; it was trained locally. This notebook only
produces `runs/stage1_v2/best.pt`.

Expected wall-clock on a T4: ~10 min setup and download, then ~1.5–2.5 h training.

## Step 0 — Mount Drive, check GPU

In [ ]:
# step 0: connect google drive, so data and checkpoints survive if colab disconnects
from google.colab import drive
drive.mount('/content/drive')

# check that a gpu is present (training on a cpu would take days)
import torch, shutil
if not torch.cuda.is_available():
    raise SystemExit('No GPU. Runtime -> Change runtime type -> T4 GPU, then re-run.')
print('GPU :', torch.cuda.get_device_name(0))
print('CUDA:', torch.version.cuda)

# check free drive space (datasets need about 4 gb, each checkpoint about 200 mb)
total, used, free = shutil.disk_usage('/content/drive/MyDrive')
print(f'Drive free: {free / 1e9:.1f} GB')
if free < 8e9:
    print('WARNING: under 8 GB free. Datasets need ~4 GB, each checkpoint ~200 MB.')

Mounted at /content/drive
GPU : Tesla T4
CUDA: 12.8
Drive free: 66.5 GB


**what happens:** drive is mounted and the gpu is checked (a Tesla T4 in this run, 66.5 GB free).

**why:** training takes about an hour, so checkpoints must be saved on drive, and a gpu is needed.

## Step 1 — Clean sync of the repo (private repo)

**One-time setup.** The repo is private, so Colab needs a token to read it:

1. Go to [github.com/settings/tokens](https://github.com/settings/tokens) → *Generate new token (classic)*
2. Tick the **`repo`** scope. Set a short expiry — 7 days is plenty.
3. Copy the token (`ghp_...`). GitHub shows it only once.
4. In Colab, click the **key icon** in the left sidebar → *Add new secret*
   - Name: `GITHUB_TOKEN`
   - Value: paste the token
   - Toggle **Notebook access** on

The token is read from Colab Secrets, never written into this notebook, never stored in
`.git/config` on your Drive, and is scrubbed from anything this cell prints. Revoke it on
GitHub once the project is done.

**Why a hard reset.** A plain `git pull` into an existing Drive copy can fail on a conflict
and leave you silently on **old code** — which would train the old taxonomy and look
completely normal while doing it. This discards local edits in the Drive copy; that copy is
a training scratch checkout, so that is what we want.

In [ ]:
import os, subprocess

# the project code is in a private github repo; this cell downloads the latest version of it
REPO_SLUG = 'darkhorse0204/edge-waste'
REPO_DIR = '/content/drive/MyDrive/edge-waste'
CLEAN_URL = f'https://github.com/{REPO_SLUG}'

# read the token. raising inside the except block makes ipython's traceback
# formatter melt down, so collect the failure and report it afterwards.
# read the github token from colab secrets (it is never typed in the notebook)
TOKEN, _why = None, None
try:
    from google.colab import userdata
    TOKEN = userdata.get('GITHUB_TOKEN')
except ImportError:
    _why = 'not running in Colab'
except Exception as exc:
    _why = {
        'SecretNotFoundError': 'the secret does not exist yet',
        'NotebookAccessError': 'the secret exists but Notebook access is OFF',
    }.get(type(exc).__name__, type(exc).__name__)

if not TOKEN:
    print('=' * 72)
    print(f'  SETUP NEEDED — GITHUB_TOKEN unavailable ({_why or "it is empty"}).')
    print('=' * 72)
    print('  1. github.com/settings/tokens -> Generate new token (classic)')
    print('     Tick the  repo  scope. 7-day expiry is fine. Copy it (ghp_...).')
    print('  2. In Colab: click the KEY icon in the far-left sidebar')
    print('     -> + Add new secret')
    print('        Name : GITHUB_TOKEN        (exactly this, case-sensitive)')
    print('        Value: paste the token')
    print('        Then toggle  Notebook access  ON  <- easy to miss')
    print('  3. Re-run this cell.')
    print('=' * 72)
    raise SystemExit('GITHUB_TOKEN not set - follow the steps printed above.') from None

# used only as a git argument; never written to .git/config on drive.
# the token is used only inside this url for git, and is never saved on drive
AUTH_URL = f'https://x-access-token:{TOKEN}@github.com/{REPO_SLUG}'


# small helper: run a shell command and hide the token in anything it prints
def sh(cmd, cwd=None, check=None):
    r = subprocess.run(cmd, shell=True, cwd=cwd, capture_output=True, text=True)
    # scrub the token from anything git echoes back before printing.
    print((r.stdout + r.stderr).replace(TOKEN, '***').strip())
    if check and r.returncode != 0:
        raise SystemExit(check)
    return r.returncode


# make the drive copy exactly equal to the latest github code (hard reset)
def sync_in_place(cwd):
    """Fetch origin/main and hard-reset to it, leaving untracked files alone.

    Untracked here means data/ and runs/ — gitignored, and potentially
    gigabytes of downloaded datasets and checkpoints. A reset touches tracked
    files only, so those survive.
    """
    sh(f'git fetch {AUTH_URL} main', cwd=cwd,
        check='Fetch failed - token may lack `repo` scope, or it expired.')
    sh('git reset --hard FETCH_HEAD', cwd=cwd)
    sh('git clean -fd src configs scripts', cwd=cwd)


# three cases: checkout exists / folder exists without git / nothing exists yet
has_dir = os.path.isdir(REPO_DIR)
has_git = os.path.isdir(os.path.join(REPO_DIR, '.git'))

if has_git:
    print('Existing checkout found, currently at:')
    sh('git log --oneline -1', cwd=REPO_DIR)
    print('\n--- syncing to origin/main ---')
    sync_in_place(REPO_DIR)

elif has_dir and os.listdir(REPO_DIR):
    # a non-empty folder with no .git: a previous clone died partway, or the
    # folder predates git. cloning into it is impossible, but deleting it
    # could throw away downloaded datasets. convert it in place instead.
    print(f'{REPO_DIR} exists but is not a git checkout. Contents:')
    for name in sorted(os.listdir(REPO_DIR))[:15]:
        print('   ', name)
    print('\n--- converting it into a checkout in place (keeps data/ and runs/) ---')
    sh('git init -q', cwd=REPO_DIR)
    sh(f'git remote add origin {CLEAN_URL}', cwd=REPO_DIR)
    sync_in_place(REPO_DIR)

else:
    print('No checkout found - cloning fresh.')
    sh(f'git clone {AUTH_URL} {REPO_DIR}',
        check='Clone failed - token may lack `repo` scope, or it expired.')
    sh(f'git remote set-url origin {CLEAN_URL}', cwd=REPO_DIR)

os.chdir(REPO_DIR)
print('\n--- now at ---')
sh('git log --oneline -3')

# confirm no token leaked into the repo config on drive.
# safety check: the token must not be stored in the git config on drive
cfg_path = os.path.join(REPO_DIR, '.git', 'config')
if os.path.exists(cfg_path):
    assert TOKEN not in open(cfg_path).read(), 'Token leaked into .git/config - remove it.'
    print('\n.git/config is clean (no token stored)')

Existing checkout found, currently at:
b1c7842 Survive corrupt images instead of losing the training run to one

--- syncing to origin/main ---
From https://github.com/darkhorse0204/edge-waste
 * branch            main       -> FETCH_HEAD
HEAD is now at 0a870a9 Restructure repo, add ML analysis suite, fix training recipe
Updating files:  78% (111/141)
Updating files:  79% (112/141)
Updating files:  80% (113/141)
Updating files:  81% (115/141)
Updating files:  82% (116/141)
Updating files:  83% (118/141)
Updating files:  84% (119/141)
Updating files:  85% (120/141)
Updating files:  86% (122/141)
Updating files:  87% (123/141)
Updating files:  88% (125/141)
Updating files:  89% (126/141)
Updating files:  90% (127/141)
Updating files:  91% (129/141)
Updating files:  92% (130/141)
Updating files:  93% (132/141)
Updating files:  94% (133/141)
Updating files:  95% (134/141)
Updating files:  96% (136/141)
Updating files:  97% (137/141)
Updating files:  98% (139/141)
Updating files:  99% (140/

**what happens:** the repo is brought to the newest version of the code.

**why a hard reset:** a normal `git pull` can fail on a conflict and silently leave old code, which would train the old 7-class taxonomy without any error.

In [ ]:
import os, shutil
os.chdir(REPO_DIR)
# delete old processed images and the old split, so stale duplicate images cannot enter the new split (this fixed the 1,234-steps bug)
shutil.rmtree('data/processed', ignore_errors=True)
if os.path.exists('data/splits.csv'): os.remove('data/splits.csv')
print(os.listdir('data'))   # should still show 'downloads'


['downloads']


**what happens:** old processed images and the old split file are deleted; only the downloads stay.

**why:** an earlier run showed 1,234 steps per epoch instead of 617, because old copies of every image were still on drive and the same photo could land in both train and test. deleting them makes the split clean again.

## Step 2 — Install the package

In [ ]:
import os, sys
os.chdir(REPO_DIR)
# install our package (edgewaste) in editable mode, so colab runs the code from the repo
!pip install -e '.[data]' -q

src = f'{REPO_DIR}/src'
if src not in sys.path:
    sys.path.insert(0, src)
print('installed')

  Installing build dependencies ... done
  Checking if build backend supports build_editable ... done
  Getting requirements to build editable ... done
  Preparing editable metadata (pyproject.toml) ... done
  Building editable for edgewaste (pyproject.toml) ... done
installed


**what happens:** `pip install -e .` makes the folder `src/edgewaste/` importable and creates the commands `edgewaste-fetch`, `edgewaste-ingest`, `edgewaste-split`, `edgewaste-train`, `edgewaste-eval` (they are listed in `pyproject.toml`).

## Step 3 — GATE: verify the code is the 33-class version

Stops if the checkout is stale or the config would train from scratch. `pretrained: false`
is checked explicitly because it is the exact flag that previously collapsed accuracy from
~95% to 49%, and it fails silently — the run just ends up bad.

In [ ]:
import importlib
import edgewaste.taxonomy as tx
import edgewaste.config as cf
importlib.reload(tx); importlib.reload(cf)

print(tx.summary())

# load the training settings from the yaml file
cfg = cf.Config.load('configs/classifier_convnext_vit.yaml')
print('\npretrained backbones :', cfg.model.pretrained)
print('epochs               :', cfg.train.epochs)
print('batch size           :', cfg.train.batch_size)

# gate 1: stop if the code is old (it must have 33 classes)
assert tx.NUM_CLASSES == 33, (
    f'Expected 33 classes, found {tx.NUM_CLASSES}. Checkout is stale — re-run Step 1.')
# gate 2: stop if pretrained weights are off (training from scratch gave only 49%)
assert cfg.model.pretrained is True, (
    'pretrained is False. Training from scratch collapses accuracy to ~49%. '
    'Fix configs/classifier_convnext_vit.yaml before continuing.')
print(f'\nGATE PASSED: {tx.NUM_CLASSES} classes, pretrained backbones enabled.')

Canonical taxonomy: 33 item classes in 9 material families

  PLASTIC  (9)
     0 plastic_water_bottles
     1 plastic_soda_bottles
     2 plastic_detergent_bottles
     3 plastic_food_containers
     4 plastic_shopping_bags
     5 plastic_trash_bags
     6 plastic_cup_lids
     7 plastic_straws
     8 disposable_plastic_cutlery

  PAPER  (4)
     9 newspaper
    10 magazines
    11 office_paper
    12 paper_cups

  CARDBOARD  (2)
    13 cardboard_boxes
    14 cardboard_packaging

  GLASS  (3)
    15 glass_beverage_bottles
    16 glass_food_jars
    17 glass_cosmetic_containers

  METAL  (4)
    18 aluminum_soda_cans
    19 aluminum_food_cans
    20 steel_food_cans
    21 aerosol_cans

  ORGANIC  (4)
    22 food_waste
    23 eggshells
    24 coffee_grounds
    25 tea_bags

  STYROFOAM  (2)
    26 styrofoam_cups
    27 styrofoam_food_containers

  TEXTILE  (2)
    28 clothing
    29 shoes

  HAZARDOUS  [HAZARDOUS]  (3)
    30 battery
    31 e_waste
    32 medical

Sources:
  recyclable_

**what happens:** the class list and the training config are loaded and two safety gates run.

**where:** classes are defined in `src/edgewaste/taxonomy.py`; settings are read by `Config.load` in `src/edgewaste/config.py` from `configs/classifier_convnext_vit.yaml`.

**why the gates:** a wrong class list or `pretrained: false` would not give an error, only a bad model.

## Step 3b — (optional) move the datasets to local disk

**Skip this unless downloads are painfully slow.** Default is Drive, and that
is usually the right call.

Drive is mounted over FUSE: fine for a few large files, slow for many small
ones. These datasets are ~20,000 individual images, so the unzip onto Drive
takes several minutes per dataset where local disk takes about one. The
download itself is the same either way — the cost is in the per-file writes.

But local disk is wiped when the runtime recycles, and training runs 1.5–2.5 h
with disconnects being common. Data on Drive survives that; data on `/content`
means re-downloading ~4 GB every reconnect.

| | Drive (default) | Local `/content` |
|---|---|---|
| unzip | several min per dataset | ~1 min |
| survives disconnect | yes | no |

So: leave it alone unless you are stuck re-running setup repeatedly in one
session. Checkpoints go to Drive either way.

In [ ]:
# leave false to keep datasets on drive (persistent across disconnects).
# set true only if the per-file unzip onto drive is the bottleneck for you.
USE_LOCAL_DATA = False

import os, yaml

if USE_LOCAL_DATA:
    LOCAL = '/content/edge_waste_data'
    os.makedirs(LOCAL, exist_ok=True)
    cfg_path = 'configs/classifier_convnext_vit.yaml'
    raw = yaml.safe_load(open(cfg_path))
    raw.setdefault('data', {}).update({
        'downloads_dir': f'{LOCAL}/downloads',
        'processed_dir': f'{LOCAL}/processed',
        'manifest': f'{LOCAL}/splits.csv',
    })
    # checkpoints stay on drive regardless — they are what must survive.
    raw.setdefault('train', {})['output_dir'] = f'{REPO_DIR}/runs/stage1_v2'
    # written to the working copy only; step 1 resets it next session.
    yaml.safe_dump(raw, open(cfg_path, 'w'), sort_keys=False)
    print(f'datasets -> {LOCAL}  (local, fast, wiped on disconnect)')
    print(f'checkpoints -> {REPO_DIR}/runs/stage1_v2  (Drive, persistent)')
else:
    print('datasets -> Drive (default: slower unzip, survives disconnects)')

datasets -> Drive (default: slower unzip, survives disconnects)


## Step 4 — Kaggle credentials

Upload your `kaggle.json` to the **root of your Drive** (`MyDrive/kaggle.json`) first.

In [ ]:
import os, shutil

# kaggle needs a key file (kaggle.json) to allow the dataset download
src_json = '/content/drive/MyDrive/kaggle.json'
if not os.path.exists(src_json):
    raise SystemExit(
        'kaggle.json not found at MyDrive/kaggle.json. Get it from kaggle.com/settings '
        '-> Create New Token, upload to your Drive root, then re-run this cell.')

os.makedirs(os.path.expanduser('~/.kaggle'), exist_ok=True)
shutil.copy(src_json, os.path.expanduser('~/.kaggle/kaggle.json'))
os.chmod(os.path.expanduser('~/.kaggle/kaggle.json'), 0o600)
print('credentials installed')
!kaggle datasets list -s garbage 2>&1 | head -3

credentials installed
ref                                                        title                                                  size  lastUpdated                 downloadCount  voteCount  usabilityRating  
---------------------------------------------------------  -----------------------------------------------  ----------  --------------------------  -------------  ---------  ---------------  


## Step 5 — Download the datasets

Three sources: the 30-class household dataset (the backbone), plus TrashBox and
Garbage-12 for the hazardous classes the household set lacks. Roughly 3–4 GB total.
Already-present sources are skipped, so this is safe to re-run.

In [ ]:
os.chdir(REPO_DIR)
# download the 3 public datasets (code: src/edgewaste/data/download_kaggle_datasets.py)
!edgewaste-fetch --config configs/classifier_convnext_vit.yaml

  [have] recyclable_household: already present at data/downloads/recyclable_household
  [have] trashbox: already present at data/downloads/trashbox
  [have] garbage12: already present at data/downloads/garbage12

Downloaded 3/3 public sources into data/downloads
Next: run `edgewaste-ingest` to merge them into canonical classes.


**what happens:** three public kaggle datasets are present (`[have]` means they were already downloaded).

**where:** `src/edgewaste/data/download_kaggle_datasets.py`; dataset names are in `src/edgewaste/taxonomy.py` (`SOURCES`).

## Step 6 — Ingest, split, and GATE on the data

Consolidates sources into canonical classes and builds the stratified split, then checks
that **all 33 classes actually have images** — a renamed folder upstream would otherwise
just produce a silently smaller taxonomy.

In [ ]:
os.chdir(REPO_DIR)
# ingest: copy images into one folder per class using the mapping in taxonomy.py; unreadable images are skipped
!edgewaste-ingest --config configs/classifier_convnext_vit.yaml
# split: stratified 70/15/15 train/val/test with seed 42 (code: src/edgewaste/data/make_data_splits.py)
!edgewaste-split  --config configs/classifier_convnext_vit.yaml

- ingesting 'recyclable_household' from data/downloads/recyclable_household
- ingesting 'trashbox' from data/downloads/trashbox
/usr/local/lib/python3.13/dist-packages/PIL/Image.py:1047: UserWarning: Palette images with Transparency expressed in bytes should be converted to RGBA images
  warnings.warn(
  [warn] trashbox: skipped 1 unreadable image(s), e.g. data/downloads/trashbox/TrashBox_train_set/e-waste/e-waste 1719.jpg
- ingesting 'garbage12' from data/downloads/garbage12

Consolidated per-class counts (processed_dir):
  plastic_water_bottles    500
  plastic_soda_bottles    500
  plastic_detergent_bottles    500
  plastic_food_containers    500
  plastic_shopping_bags    500
  plastic_trash_bags    500
  plastic_cup_lids    500
  plastic_straws    500
  disposable_plastic_cutlery    500
  newspaper         500
  magazines         500
  office_paper      500
  paper_cups        500
  cardboard_boxes    500
  cardboard_packaging    500
  glass_beverage_bottles    500
  glass_food_ja

**what happens:** images from the 3 sources are mapped to the 33 classes, one unreadable image is skipped, then the 70/15/15 split is written to `data/splits.csv`.

**where:** `ingest_source` in `src/edgewaste/data/ingest_raw_datasets.py` and `main` in `src/edgewaste/data/make_data_splits.py`.

**why stratified:** every class keeps the same share in train, validation and test, so rare classes are tested too.

In [ ]:
import pandas as pd
import edgewaste.taxonomy as tx

# read the split file and count images per class
df = pd.read_csv('data/splits.csv')
names = list(tx.CLASS_NAMES)
df['class_name'] = df['label'].map(dict(enumerate(names)))
counts = df.groupby('class_name').size().reindex(names, fill_value=0)

print(f'total images: {len(df)}\n')
for fam in tx.FAMILIES:
    members = tx.FAMILY_TO_CLASSES[fam]
    print(f'{fam.upper():<12} {counts[list(members)].sum():>6}')
    for m in members:
        print(f'   {m:<32} {counts[m]:>6}')

print('\nper-split totals:')
print(df.groupby('split').size().to_string())

# gate 4: every one of the 33 classes must have images
empty = [c for c in names if counts[c] == 0]
assert not empty, (
    f'{len(empty)} classes have NO images: {empty}. A source folder was probably renamed '
    'upstream — fix the mapping in taxonomy.py before training.')
thin = [c for c in names if 0 < counts[c] < 50]
if thin:
    print(f'\nWARNING: very few images for {thin} — expect weak recall there.')
print(f'\nGATE PASSED: all {len(names)} classes populated.')

total images: 28202

PLASTIC        4500
   plastic_water_bottles               500
   plastic_soda_bottles                500
   plastic_detergent_bottles           500
   plastic_food_containers             500
   plastic_shopping_bags               500
   plastic_trash_bags                  500
   plastic_cup_lids                    500
   plastic_straws                      500
   disposable_plastic_cutlery          500
PAPER          2000
   newspaper                           500
   magazines                           500
   office_paper                        500
   paper_cups                          500
CARDBOARD      1000
   cardboard_boxes                     500
   cardboard_packaging                 500
GLASS          1500
   glass_beverage_bottles              500
   glass_food_jars                     500
   glass_cosmetic_containers           500
METAL          2000
   aluminum_soda_cans                  500
   aluminum_food_cans                  500
   steel_food_cans 

**what happens:** the split is counted per family and class, and three checks run (no stale duplicates, all 33 classes present, no class with fewer than 50 images).

**why:** these problems do not raise an error by themselves; they only make the model quietly worse.

## Step 7 — Smoke test (~2 min)

One epoch on a 64-image subset. Proves the whole path works before committing hours to it.

In [ ]:
os.chdir(REPO_DIR)
# smoke test: 1 epoch on a tiny subset, only to check that everything is connected
!edgewaste-train --config configs/classifier_convnext_vit.yaml --smoke
print('\nsmoke test passed — pipeline is wired correctly')

Device: cuda
Classes present in train: 33/33
Unexpected keys (norm.bias, norm.weight) found while loading pretrained weights. This may be expected if model is being adapted.
Trainable parameters: 50,752,002
epoch 1: backbones frozen (linear-probe phase)
epoch  1 | train_loss 3.349 acc 0.500 | val_loss 3.203 acc 1.000 | 9.5s  *

Best val acc: 1.000
Checkpoints in runs/stage1_v2

smoke test passed — pipeline is wired correctly


**what happens:** one tiny epoch. the values (accuracy 1.0 on a few images) mean nothing; the aim is only to prove that data, model, loss, optimiser and saving all work. note `Trainable parameters: 50,752,002` (the model size).

## Step 8 — Full training

~1.5–2.5 h on a T4. Checkpoints to `runs/stage1_v2/best.pt` on Drive, so a disconnect does
not lose finished epochs.

Keep this tab visible — Colab reclaims idle runtimes.

In [ ]:
os.chdir(REPO_DIR)
# the real training: runs train() in src/edgewaste/classification/train_classifier.py with the improved recipe from the yaml file
!edgewaste-train --config configs/classifier_convnext_vit.yaml

Device: cuda
/usr/local/lib/python3.13/dist-packages/torch/utils/data/dataloader.py:424: UserWarning: This DataLoader will create 4 worker processes in total. Our suggested max number of worker in current system is 2, which is smaller than what this DataLoader is going to create. Please be aware that excessive worker creation might get DataLoader running slow or even freeze, lower the worker number to avoid potential slowness/freeze if necessary.
  self.check_worker_number_rationality()
Classes present in train: 33/33
Unexpected keys (norm.bias, norm.weight) found while loading pretrained weights. This may be expected if model is being adapted.
Trainable parameters: 50,752,002
epoch 1: backbones frozen (linear-probe phase)
epoch 1/25:   0% 0/617 [00:00<?, ?it/s]/usr/local/lib/python3.13/dist-packages/torch/utils/data/dataloader.py:432: UserWarning: This DataLoader will create 4 worker processes in total. Our suggested max number of worker in current system is 2, which is smaller than w

**what happens:** the real training, with these ideas from `configs/classifier_convnext_vit.yaml`:

| idea | where in `src/edgewaste/classification/train_classifier.py` |
|---|---|
| balanced batches (one imbalance correction) | `WeightedRandomSampler`, line 53 |
| cross-entropy with label smoothing 0.1 | `nn.CrossEntropyLoss(... label_smoothing ...)`, line 105 |
| backbones learn 10x slower than new layers | two parameter groups, lines 108-110 |
| AdamW optimiser | line 111 |
| one-cycle learning rate (warm-up then slow decay) | `OneCycleLR`, line 116 |
| mixed precision (faster on the gpu) | `GradScaler` and `autocast`, lines 122 and 143 |
| backbones frozen for the first 2 epochs | `frozen = epoch <= freeze_backbone_epochs`, line 131 |
| gradient clipping at 1.0 | `clip_grad_norm_`, line 149 |
| keep the best model, early stopping | best validation accuracy is saved to `best.pt`; stop after 5 epochs without improvement |

**what the log shows:** validation accuracy is already 0.900 after epoch 1 (only the new layers trained). it is best at epoch 8 (0.939), and training stopped after epoch 13 because there was no improvement for 5 epochs. training accuracy keeps rising (0.975) while validation stays near 0.938, a small gap, which is mild overfitting but not harmful.

## Step 9 — Evaluate (item level *and* family level)

Reports both levels of the taxonomy. The gap between them is the number worth quoting:
high family accuracy with lower item accuracy means the errors sit *inside* material
families, so items would still be routed to the right stream. Hazardous recall is
reported separately.

In [ ]:
os.chdir(REPO_DIR)
# score the best checkpoint on the test split: item level, family level and hazard recall (code: src/edgewaste/classification/evaluate_classifier.py)
!edgewaste-eval --config configs/classifier_convnext_vit.yaml --ckpt runs/stage1_v2/best.pt

Unexpected keys (norm.bias, norm.weight) found while loading pretrained weights. This may be expected if model is being adapted.
/usr/local/lib/python3.13/dist-packages/torch/utils/data/dataloader.py:424: UserWarning: This DataLoader will create 4 worker processes in total. Our suggested max number of worker in current system is 2, which is smaller than what this DataLoader is going to create. Please be aware that excessive worker creation might get DataLoader running slow or even freeze, lower the worker number to avoid potential slowness/freeze if necessary.
  self.check_worker_number_rationality()
/usr/local/lib/python3.13/dist-packages/torch/utils/data/dataloader.py:432: UserWarning: This DataLoader will create 4 worker processes in total. Our suggested max number of worker in current system is 2, which is smaller than what this DataLoader is going to create. Please be aware that excessive worker creation might get DataLoader running slow or even freeze, lower the worker number to 

**what happens:** the best checkpoint is scored on the 4,232 test images.

**where:** `evaluate` in `src/edgewaste/classification/evaluate_classifier.py`; family level in `_family_level` (line 141); hazard recall in `_hazard_recall` (line 166).

**how to read the result:** item accuracy 0.9383 (exact class), family accuracy 0.9728 (correct bin), hazard recall 0.9837 (726 of 738 dangerous items were called dangerous). the 12 missed hazards are listed with where they went (mostly office paper).

**note:** the warnings about `DataLoader` workers and `Palette images` are harmless; they do not change results.

In [ ]:
from IPython.display import Image as IPImage, display
import pathlib

# show the two confusion matrices saved by the evaluation
for p in ['runs/stage1_v2/confusion_test.png', 'runs/stage1_v2/confusion_family_test.png']:
    if pathlib.Path(p).exists():
        print(p)
        display(IPImage(p))

## Step 10 — Summary and files to bring back

Download `runs/stage1_v2/best.pt` from Drive and drop it into the local repo at the same
path to run the live demo and the video survey pipeline.

In [ ]:
import json, pathlib

# read the saved metrics file and print the three headline numbers
m_path = pathlib.Path('runs/stage1_v2/metrics_test.json')
if m_path.exists():
    m = json.loads(m_path.read_text())
    item = m.get('item_level', {})
    fam = m.get('family_level', {})
    haz = m.get('hazardous', {})
    print('=== TEST RESULTS ===')
    print(f"item-level   accuracy : {item.get('accuracy', 0)*100:.2f}%  "
          f"({item.get('num_classes', '?')} classes)")
    print(f"family-level accuracy : {fam.get('accuracy', 0)*100:.2f}%  "
          f"({len(fam.get('families_present', []))} families)")
    if haz.get('support'):
        print(f"hazardous recall      : {haz['recall']*100:.2f}%  "
              f"({haz['caught']}/{haz['support']})")
    print(f"test images           : {m.get('num_samples')}")

# list the files to download from drive
print('\n=== FILES TO DOWNLOAD ===')
for f in ['runs/stage1_v2/best.pt', 'runs/stage1_v2/history.json',
          'runs/stage1_v2/metrics_test.json', 'runs/stage1_v2/confusion_test.png',
          'runs/stage1_v2/confusion_family_test.png']:
    p = pathlib.Path(f)
    print(f'  {"OK " if p.exists() else "-- "} {f}'
          + (f'  ({p.stat().st_size/1e6:.1f} MB)' if p.exists() else ''))

=== TEST RESULTS ===
item-level   accuracy : 93.83%  (33 classes)
family-level accuracy : 97.28%  (9 families)
hazardous recall      : 98.37%  (726/738)
test images           : 4232

=== FILES TO DOWNLOAD ===
  OK  runs/stage1_v2/best.pt  (203.1 MB)
  OK  runs/stage1_v2/history.json  (0.0 MB)
  OK  runs/stage1_v2/metrics_test.json  (0.0 MB)
  OK  runs/stage1_v2/confusion_test.png  (0.2 MB)
  OK  runs/stage1_v2/confusion_family_test.png  (0.0 MB)


**what happens:** the saved metrics are printed again as the final summary: **93.83% item, 97.28% family, 98.37% hazard recall** on 4,232 test images, measured on colab's own split.